File: dataset_reduzido/avatar/Python/Code/codeforces_216_B.py
def readInts ( ) : return map ( int , input ( ).split ( ) )
def dfs ( g , seen , i ) :
    if i in seen :
        return 0 , 0
    seen.add ( i )
    nodes = 1
    edges = len ( g [ i ] )
    for j in g [ i ] :
        x , y = dfs ( g , seen , j )
        nodes += x
        edges += y
    return nodes , edges
def solve ( ) :
    line0 = [ ]
    try : line0 = readInts ( )
    except EOFError : return False
    n , m = line0
    g = { }
    seen = set ( )
    for i in range ( 1 , n + 1 ) :
        g [ i ] = set ( )
    for _ in range ( m ) :
        a , b = readInts ( )
        g [ a ].add ( b )
        g [ b ].add ( a )
    ans = 0
    for i in range ( 1 , n + 1 ) :
        if i not in seen :
            nodes , edges = dfs ( g , seen , i )
            if nodes > 1 and nodes % 2 == 1 and 2 * nodes == edges :
                ans += 1
    if ( n - ans ) % 2 == 1 :
        ans += 1
    print ( ans )
    return True
while solve ( ) : pass

File: output/Qwen_repaired_1/avatar/Python/Java/codeforces_216_B.java
import java.util.*;

public class codeforces_216_B {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        while (solve(scanner)) {
        }
    }

    public static boolean solve(Scanner scanner) {
        List<Integer> line0 = new ArrayList<>();
        try {
            String[] input = scanner.nextLine().split(" ");
            for (String s : input) {
                line0.add(Integer.parseInt(s));
            }
        } catch (NoSuchElementException e) {
            return false;
        }
        int n = line0.get(0);
        int m = line0.get(1);
        Map<Integer, Set<Integer>> g = new HashMap<>();
        Set<Integer> seen = new HashSet<>();

        for (int i = 1; i <= n; i++) {
            g.put(i, new HashSet<>());
        }

        for (int i = 0; i < m; i++) {
            String[] split = scanner.nextLine().split(" ");
            int a = Integer.parseInt(split[0]);
            int b = Integer.parseInt(split[1]);
            g.get(a).add(b);
            g.get(b).add(a);
        }

        int ans = 0;
        for (int i = 1; i <= n; i++) {
            if (!seen.contains(i)) {
                Set<Integer> cycleNodes = new HashSet<>();
                int edges = dfs(g, seen, cycleNodes, i);
                if (cycleNodes.size() > 1 && cycleNodes.size() % 2 == 1 && 2 * cycleNodes.size() == edges) {
                    ans++;
                }
            }
        }

        if ((n - ans) % 2 == 1) {
            ans++;
        }

        System.out.println(ans);
        return true;
    }

    private static int dfs(Map<Integer, Set<Integer>> g, Set<Integer> seen, Set<Integer> cycleNodes, int i) {
        if (seen.contains(i)) {
            return 0;
        }
        seen.add(i);
        cycleNodes.add(i);

        int edges = 0;
        for (int neighbor : g.get(i)) {
            edges += dfs(g, seen, cycleNodes, neighbor);
        }

        return edges + cycleNodes.size() - 1;
    }
}